File: spotifyOAuth.js
export async function spotifyOAuth(){
    console.log("Authorizing")
    //Code Verifier Template from API Docs
    const generateRandomString = (length) => {
        const possible = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
        const values = crypto.getRandomValues(new Uint8Array(length));
        return values.reduce((acc, x) => acc + possible[x % possible.length], "");
    }

    //Conversion to SHA256 Template from API Docs
    const sha256 = async (plain) => {
    const encoder = new TextEncoder()
    const data = encoder.encode(plain)
    return window.crypto.subtle.digest('SHA-256', data)
    }

    //Base64encode template from API Docs
    const base64encode = (input) => {
        return btoa(String.fromCharCode(...new Uint8Array(input)))
        .replace(/=/g, '')
        .replace(/\+/g, '-')
        .replace(/\//g, '_');
    }

    const clientId = 'a9133a4606a74919aefcc1c43f11247e';
    const redirectUri = 'http://localhost:5500';

    //const scope = 'user-read-private user-read-email playlist-modify-private playlist-modify-public scope user-library-modify';
    const scope = 'user-library-modify';
    const authUrl = new URL("https://accounts.spotify.com/authorize")

    const codeVerifier  = generateRandomString(64);
    const hashed = await sha256(codeVerifier)
    const codeChallenge = base64encode(hashed);

    window.localStorage.setItem('code_verifier', codeVerifier);

    const params =  {
        response_type: 'code',
        client_id: clientId,
        scope,
        code_challenge_method: 'S256',
        code_challenge: codeChallenge,
        redirect_uri: redirectUri,
    }

    authUrl.search = new URLSearchParams(params).toString();
    console.log(authUrl.toString())
    window.location.href = authUrl.toString();

    const urlParams = new URLSearchParams(window.location.search);
    let code = urlParams.get('code');
    
    const getToken = async code => {
        
        // stored in the previous step
        let codeVerifier = localStorage.getItem('code_verifier');
    
        const payload = {
        method: 'POST',
        headers: {
            'Content-Type': 'application/x-www-form-urlencoded',
        },
        body: new URLSearchParams({
            client_id: clientId,
            grant_type: 'authorization_code',
            code,
            redirect_uri: redirectUri,
            code_verifier: codeVerifier,
        }),
        }
    
        const body = await fetch(url, payload);
        const response =await body.json();
    
        localStorage.setItem('access_token', response.access_token.json());
        console.log(response.access_token.json())
        return response.access_token.json()
    }

}

File: spotifyOAuth.py
from flask import Flask, redirect, request, session, url_for, jsonify
import requests

app = Flask(__name__)


CLIENT_ID = 'a9133a4606a74919aefcc1c43f11247e'
CLIENT_SECRET = '309db053956c4d32ad3ad229c96ac24c'
REDIRECT_URI = 'http://localhost:5000/callback'
TOKEN_URL = 'https://accounts.spotify.com/api/token'
SCOPES = 'user-read-private user-read-email'


@app.route('/login')
def login():

    auth_url = (
        f"https://accounts.spotify.com/authorize"
        f"?client_id={CLIENT_ID}"
        f"&response_type=code"
        f"&redirect_uri={REDIRECT_URI}"
        f"&scope={SCOPES}"
        f"&show_dialog=true"
    )
    
    return redirect(auth_url)

@app.route('/callback')
def callback():

    if 'error' in request.args:
        return jsonify({"error":request.args["error"]})

    if 'code' in request.args:
        req_body = {
            'code' : request.args['code'], 
            'grant_type' : 'authorization_code', 
            'redirect_uri' : REDIRECT_URI, 
            'client_id' : CLIENT_ID, 
            'client_secrect' : CLIENT_SECRET 
        }
    
    response = request.post(TOKEN_URL,data=req_body)
    token_info = response.json()

    session['access_token'] = token_info["access_token"]
    session['refresh_token'] = token_info["refresh_token"]
    

    return redirect('/add')


@app.route('/add')
def add():
    access_token = session.get('access_token')
    if not access_token:
        return redirect('/login')

    headers = {
        'Authorization': f'Bearer {access_token}'
    }

    response = requests.get('https://api.spotify.com/v1/me/playlists', headers=headers)
    
    if response.status_code != 200:
        return 'Error: Failed to fetch profile.'

    user_data = response.json()
    return jsonify(user_data)

if __name__ == '__main__':
    app.run(debug=True)